File: claro-full/frontend/wasm/engine.c
// engine.c - Minimal C reasoning engine for CLARO

#include <emscripten.h>
#include <string.h>
#include <stdlib.h>

// Simple helper to check if a substring exists
int contains(const char* haystack, const char* needle) {
  return strstr(haystack, needle) != NULL;
}

// Convert to lowercase (very basic, ASCII only)
void to_lower(char* s) {
  for (int i = 0; s[i]; i++) {
    if (s[i] >= 'A' && s[i] <= 'Z') s[i] += 32;
  }
}

// EMSCRIPTEN_KEEP ensures the function is exported
EMSCRIPTEN_KEEPALIVE
const char* reason(const char* question_json, const char* knowledge_json) {
  // For now, just a minimal demo:
  // If question contains "normalization", return a fixed JSON result.
  // In a real version, you'd parse JSON, search knowledge, etc.

  static char result[4096];

  char q[1024];
  strncpy(q, question_json, sizeof(q) - 1);
  q[sizeof(q) - 1] = '\0';
  to_lower(q);

  if (contains(q, "normalization")) {
    strcpy(result,
      "{"
      "\"status\":\"success\","
      "\"confidence\":\"high\","
      "\"sourceType\":\"Definition\","
      "\"answer\":{"
        "\"summary\":\"Normalization organizes data to reduce redundancy and improve integrity.\","
        "\"keyPoints\":[\"Reduces duplicate data\",\"Improves data integrity\",\"Uses normal forms\"],"
        "\"steps\":[\"Identify repeating groups\",\"Create separate tables\",\"Define keys\"],"
        "\"example\":\"Store customer details once and reference by ID.\","
        "\"whyItMatters\":\"Easier maintenance and fewer update anomalies.\""
      "},"
      "\"meta\":{\"subject\":\"DBMS\",\"chapter\":\"Normalization\",\"topic\":\"Introduction\"}"
      "}"
    );
  } else {
    strcpy(result,
      "{"
      "\"status\":\"nomatch\","
      "\"confidence\":\"low\","
      "\"sourceType\":\"Estimate\","
      "\"answer\":null,"
      "\"meta\":null"
      "}"
    );
  }

  return result;
}

File: claro-full/frontend/wasm/engine.js
// wasm/engine.js
// Temporary JS implementation of the reasoning engine.
// Later, this will be replaced by a C → WebAssembly module.

export function reason(question, knowledgeBase) {
  const q = (question || '').toLowerCase();

  // Example: very small rule-based reasoning for demo subjects.
  // This is intentionally simple and transparent.

  // 1) DBMS normalization concept
  if (/(what is|define|explain)\s+normaliz/.test(q)) {
    const entry = knowledgeBase.find(k =>
      /normaliz/i.test(k.question) && /dbms|database/i.test(k.subject || '')
    );
    if (entry) {
      return {
        status: 'success',
        confidence: 'high',
        sourceType: entry.sourceType || 'Definition',
        answer: entry.answer,
        meta: {
          subject: entry.subject || 'DBMS',
          chapter: entry.chapter || 'Normalization',
          topic: entry.topic || 'Introduction'
        }
      };
    }
  }

  // 2) Newton's second law
  if (/(newton.*second|f\s*=\s*m\s*[\*x]\s*a|force.*mass.*acceleration)/.test(q)) {
    const entry = knowledgeBase.find(k =>
      /newton.*second|F = m|force.*mass/i.test(k.question)
    );
    if (entry) {
      return {
        status: 'success',
        confidence: 'medium',
        sourceType: entry.sourceType || 'Concept',
        answer: entry.answer,
        meta: {
          subject: entry.subject || 'Physics',
          chapter: entry.chapter || 'Laws of Motion',
          topic: entry.topic || "Newton's Second Law"
        }
      };
    }
  }

  // 3) Ohm's law
  if (/(ohm.*law|v\s*=\s*i\s*[\*x]\s*r|voltage.*current.*resistance)/.test(q)) {
    const entry = knowledgeBase.find(k =>
      /ohm/i.test(k.question)
    );
    if (entry) {
      return {
        status: 'success',
        confidence: 'medium',
        sourceType: entry.sourceType || 'Formula',
        answer: entry.answer,
        meta: {
          subject: entry.subject || 'Physics',
          chapter: entry.chapter || 'Current Electricity',
          topic: entry.topic || "Ohm's Law"
        }
      };
    }
  }

  // 4) Operating system basic
  if (/(what is|define|explain)\s+(an\s+)?operating\s+system/i.test(q)) {
    const entry = knowledgeBase.find(k =>
      /operating system|OS/i.test(k.question)
    );
    if (entry) {
      return {
        status: 'success',
        confidence: 'medium',
        sourceType: entry.sourceType || 'Definition',
        answer: entry.answer,
        meta: {
          subject: entry.subject || 'Operating Systems',
          chapter: entry.chapter || 'Introduction',
          topic: entry.topic || 'OS Basics'
        }
      };
    }
  }

  // If nothing matches strongly, return nomatch
  return {
    status: 'nomatch',
    confidence: 'low',
    sourceType: 'Estimate',
    answer: null,
    meta: null
  };
}